loader/src/main/java/net/neoforged/fml/ModLoader.java

```diff
@@ -190,22 +191,57 @@ public static void waitForTask(String name, Runnable periodicTask, CompletableFu
         }
     }
 
+    /**
+     * Exception that is fired when a mod loading future cannot be executed because a dependent future failed.
+     * It is only used for control flow and easy filtering out, but never logged or propagated further.
+     */
+    private static class DependentFutureFailedException extends RuntimeException {}
+
     /**
      * Dispatches a task across all mod containers in parallel, with progress displayed on the loading screen.
      */
     public static void dispatchParallelTask(String name, Executor parallelExecutor, Runnable periodicTask, Consumer<ModContainer> task) {
         var progress = StartupNotificationManager.addProgressBar(name, modList.size());
         try {
             periodicTask.run();
+            Map<IModInfo, CompletableFuture<Void>> modFutures = new IdentityHashMap<>();
             var futureList = modList.getSortedMods().stream()
                     .map(modContainer -> {
-                        return CompletableFuture.runAsync(() -> {
-                            ModLoadingContext.get().setActiveContainer(modContainer);
-                            task.accept(modContainer);
-                        }, parallelExecutor).whenComplete((result, exception) -> {
-                            progress.increment();
-                            ModLoadingContext.get().setActiveContainer(null);
-                        });
+                        // Collect futures for all dependencies first
+                        var depFutures = LoadingModList.get().getDependencies(modContainer.getModInfo()).stream()
+                                .map(modInfo -> {
+                                    var future = modFutures.get(modInfo);
+                                    if (future == null) {
+                                        throw new IllegalStateException("Dependency future for mod %s which is a dependency of %s not found!".formatted(
+                                                modInfo.getModId(), modContainer.getModId()));
+                                    }
+                                    return future;
+                                })
+                                .toArray(CompletableFuture[]::new);
+
+                        // Build the future for this container
+                        var future = CompletableFuture.allOf(depFutures)
+                                .<Void>handleAsync((void_, exception) -> {
+                                    if (exception != null) {
+                                        // If there was any exception, short circuit.
+                                        // The exception will already be handled by `waitForFuture` since it comes from another mod.
+                                        LOGGER.error("Skipping {} future for mod {} because a dependency threw an exception.", name, modContainer.getModId());
```

Comment:
So, there are two alternatives here (behavior wise):
- We just catch the error and let dependent tasks execute normally. This can lead to errors in the dependent task, and seems quite pointless anyway.
- We let the error propagate via `.thenRunAsync`. This will lead to the same "issue" being displayed multiple times.

So the best course of action for us is to ignore the error. But I wanted to make it clear that we are explicitly bypassing the mod, rather than just silently swallowing the error.